src/Microsoft.AspNet.Mvc.TagHelpers/Internal/GlobbingUrlBuilder.cs

```diff
@@ -75,61 +82,76 @@ public GlobbingUrlBuilder(IFileProvider fileProvider, IMemoryCache cache, PathSt
         /// <returns>The list of URLs</returns>
         public virtual ICollection<string> BuildUrlList(string staticUrl, string includePattern, string excludePattern)
         {
-            var urls = new HashSet<string>(StringComparer.Ordinal);
 
-            // Add the statically declared url if present
+            var urls = new List<string>();
             if (staticUrl != null)
             {
                 urls.Add(staticUrl);
             }
 
             // Add urls that match the globbing patterns specified
-            var matchedUrls = ExpandGlobbedUrl(includePattern, excludePattern);
-            urls.UnionWith(matchedUrls);
+            var globbedUrls = ExpandGlobbedUrl(includePattern, excludePattern);
+
+            // ExpandGlobbedUrl returns a sorted list. We can take advantage of this to perform an order preserving distinct.
+            string lastAddedEntry = null;
+            for (var i = 0; i < globbedUrls.Count; i++)
+            {
+                if (!string.Equals(staticUrl, globbedUrls[i], StringComparison.Ordinal) &&
+                    !string.Equals(lastAddedEntry, globbedUrls[i], StringComparison.Ordinal))
+                {
+                    lastAddedEntry = globbedUrls[i];
+                    urls.Add(lastAddedEntry);
+                }
+            }
 
             return urls;
         }
 
-        private IEnumerable<string> ExpandGlobbedUrl(string include, string exclude)
+        private IList<string> ExpandGlobbedUrl(string include, string exclude)
         {
             if (string.IsNullOrEmpty(include))
             {
-                return Enumerable.Empty<string>();
+                return EmptyList;
             }
-
-            var includePatterns = include.Split(PatternSeparator, StringSplitOptions.RemoveEmptyEntries);
-            var excludePatterns = exclude?.Split(PatternSeparator, StringSplitOptions.RemoveEmptyEntries);
-
-            if (includePatterns.Length == 0)
+            else if (Cache == null)
             {
-                return Enumerable.Empty<string>();
-            }
+                var includePatterns = include.Split(PatternSeparator, StringSplitOptions.RemoveEmptyEntries);
+                if (includePatterns.Length == 0)
+                {
+                    return EmptyList;
+                }
 
-            if (Cache != null)
+                return FindFiles(includePatterns, exclude);
+            }
+            else
             {
-                var cacheKey = $"{nameof(GlobbingUrlBuilder)}-inc:{include}-exc:{exclude}";
-                IEnumerable<string> files;
+                var cacheKey = new GlobbingUrlKey(include, exclude);
```

Comment:
The unique id doesn't capture the value of patterns which are the things to vary on